xrpl/clients/websocket_client.py

```diff
@@ -0,0 +1,119 @@
+"""A client for interacting with the rippled JSON RPC."""
+from __future__ import annotations
+
+import asyncio
+import json
+from typing import Callable
+
+import websockets
+
+from xrpl.clients.client import Client
+from xrpl.clients.utils import request_to_websocket, websocket_to_response
+from xrpl.models.requests.request import Request
+from xrpl.models.response import Response
+
+
+class WebsocketClient(Client):
+    """A client for interacting with the rippled WebSocket API."""
+
+    def __init__(self: WebsocketClient, url: str) -> None:
+        """
+        Constructs a WebsocketClient.
+
+        Arguments:
+            url: The URL of the rippled node to submit requests to.
+        """
+        self.url = url
+        self.websocket = None
+
+    async def request_async(self: WebsocketClient, request_object: Request) -> Response:
+        """
+        Asynchronously submit the request represented by the request_object to the
+        rippled node specified by this client's URL.
+
+        Arguments:
+            request_object: An object representing information about a rippled request.
+
+        Returns:
+            The response from the server, as a Response object.
+        """
+        formatted_request = request_to_websocket(request_object)
+        if "id" not in formatted_request:
+            formatted_request["id"] = "request_{}".format(formatted_request["command"])
+        async with websockets.connect(self.url) as websocket:
+            await websocket.send(json.dumps(formatted_request))
+            response = await websocket.recv()
+            response_dict = json.loads(response)
+            assert response_dict["id"] == formatted_request["id"]
```

Comment:
we should be raising a custom error here, IMO, instead of just an `AssertionError`

Reply:
Whoops, meant to do that, missed it